### DISCUSSION TITLE
Setting page caching for SSR

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#181] [Proposal]
## Motivation

Cache headers are notoriously difficult to understand. While we should provide low-level access to modify response headers, it would also be nice for a higher-level API for setting cache headers.

## Comparison

SvelteKit [provides](https://kit.svelte.dev/docs/loading) a `cache` object that can be returned from the `load` function that looks like this:

```js
cache: {
  "maxage": 300,
  "private": false
}
```

Next.js provides a `revalidate` option to set the number of seconds until the document should be refreshed.

## Proposal

A `Astro.cache()` method that can be used in frontmatter to set cache strategies for the current page. The function would be available on page and non-page components, enabling you to set caching in a higher-order component (such as in a layout component).

The following common strategies will be supported. You can use more than on at a time (which is common).

### stale-while-revalidate

Use the stale-while-revalidate strategy which serves a cached version for a number of seconds, while revalidating in the background.

```astro
---
Astro.cache({
  revalidate: 200
});
---
<html>
  ...
<html>
```

## maxAge

Set the max-age cache-control setting, signifying the number of seconds to cache.

```astro
---
Astro.cache({
  maxAge: 200
});
---
<html>
  ...
<html>
```

### public / private

Signifies where a public or private cache should be used. Private caches is for logged in user data.

```astro
---
Astro.cache({
  private: true
});
---
<html>
  ...
<html>
```

### immutable

```astro
---
Astro.cache({
  immutable: true
});
---
<html>
  ...
<html>
```

## Alternatives

The API `Astro.cache` could also be named:

- `Astro.setCache`
- `Astro.setCacheStrategies`

Additionally, we could forgo this API entirely and have people set headers manually:

```astro
---
Astro.response.headers.set('Cache-Control', 'max-age=604800');
---
<html>
  ...
<html>
```

**Comment by MedLeon** [comment]
At first glance I thought this would be be a Cache abstraction on the server. Astro.cache() lead me towards the wrong direction while skimming the page. 

Personally I don't think that a higher-level API is needed. But if implemented I personally would recommend going with: 
`Astro.response.headers.setCache({maxAge: 200, ...})`

This would make it 100% clear, that this is a SSR thing and you are only changing the response header and no caching is performed on the server. There even could be an easy error output if no Adapter is mentioned.

**Reply by readonlychild** [reply]
I also had a mental idea that Astro SSR would do server-side caching until I read the last piece involving `headers`.

mental idea: Oh, Astro will run this logic and I can tell it to return the same thing for the next 20 minutes without hitting the database again...

**Reply by matthewp** [reply]
If you have a CDN in front of SSR, most of those respect cache headers and indeed would not call your origin server/function for 20 minutes (or whatever you set).

**Reply by bholmesdev** [reply]
I don't think we should augment the web-standard `headers` object with a `setCache` function. The goal is the keep `request` and `response` 100% stock, with helpers going on the `Astro` global only. I personally don't see a problem with `Astro.setCache(...)` over `Astro.response.headers.setCache(...)` but it could come down to preference!

**Comment by dearcodes** [comment]
I don't think this abstraction is needed. Exposing `Astro.response.headers` being an instance of [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers/Headers) should give enough flexibility to support all `Cache-Control` strategies.

**Reply by jamesarosen** [reply]
How would you set the headers on generated `.css` and `.js` files? There aren't routes for those, so I don't know where I would put code to modify `Astro.response.headers`.

**Reply by MedLeon** [reply]
The hosting platform you are building on (Cloudflare, Vercel, Netlify, Node, Deno) most likely has an API for that. (For example in cloudflare you can deploy a _headers file that defines them).
There isn't any abstraction in Astro yet as far as i know.

**Reply by jamesarosen** [reply]
They sure do! Proposals like https://github.com/withastro/roadmap/discussions/174 would add some standardization to the Astro community.

**Reply by jamesarosen** [reply]
> The hosting platform you are building on (Cloudflare, Vercel, Netlify, Node, Deno) most likely has an API for that.

Following up on this: the Node adapter doesn't have any built-in way to do this. The best option there is probably to use it in middleware mode and add a custom Fastify / Hono / Express server around it.

**Reply by mtimbs** [reply]
It's also a blunt instrument. Eg the _headers file for cloudflare is a static file. I can't dynamically determine at render time whether to set cache control headers. Just exposing some way to return arbitrary headers with an Astro page would allow me to cache a particular page on a per route basis.

**Comment by eighty4** [comment]
Is there a cache abstraction for SSR? Curious if it's possible to have a page rendered only once and reserved from the cache + having an eTAG resolved without re-rendering.

**Reply by oemer-aran** [reply]
Personally, this is one of the few things missing for me to really ditch Next.js. Revalidation would be such a huge improvement too, as many CMS have API request limits.

**Reply by bholmesdev** [reply]
Hey @eighty4 and @oemer-aran, can you say more about expectations for a cache abstraction? I see there were a few upvoted comments that felt `Astro.response.headers` offered enough control, which steered us away from implementing this feature.

**Reply by oemer-aran** [reply]
@bholmesdev You are right. An abstraction might not be neccessary. Maybe a dedicated part for ISR and On-Demand Revalidation (independently of Vercel) is what I was actually searching for. Maybe a bit unrelated here.

I saw there were already some efforts to add documentation here https://github.com/withastro/docs/pull/3527, but the PR was closed. ISR + On-Demand Revalidation would be really helpful to serve live data from a CMS that have a API request limit, like Sanity.

**Reply by eighty4** [reply]
Caching on the server was what I asked about, where cache control headers is specifying how to control the client's browser cache of the web resources.

Web caching with cache control headers work great when you have a page that won't change or change frequently and can be cached on a client for an extended amount of time. eTAGs work for requesting a resource and knowing whether or not the client needs updated HTML based on the eTAG sent to the server. The caching I had in mind was for an SSR build that could be sent to any number of clients, so even if the client hasn't requested the page yet, and doesn't have it cached, when requesting the page, the HTML/CSS/JS is already built and sitting in memory.

**Comment by oemer-aran** [comment]
Is there any progress on this topic? This would be awesome!

**Comment by ascorbic** [comment]
I made a library as a generic abstraction to help with this. The README includes an example using Astro. https://github.com/ascorbic/cdn-cache-control

**Comment by ascorbic** [comment]
One thing we might want to do is implement optional built-in caching in the Node adapter. Hosts with their own CDN are able to use their built-in caching, but self-hosted node might benefit from an ability to do server caching. I'm assuming there are middleware modules out there that are designed to handle this sort of thing.

**Reply by florian-lefebvre** [reply]
@ascorbic should we close this in favor of #1140?

**Reply by ascorbic** [reply]
Yes, I forgot about this one when I created the new discussion
